src/Middleware/HttpLogging/src/HttpLoggingMiddleware.cs

```diff
@@ -0,0 +1,295 @@
+                if (_options.LoggingFields.HasFlag(HttpLoggingFields.Path))
+                {
+                    AddToList(list, nameof(request.PathBase), request.PathBase);
+                    AddToList(list, nameof(request.Path), request.Path);
+                }
+
+                if (_options.LoggingFields.HasFlag(HttpLoggingFields.Query))
+                {
+                    AddToList(list, nameof(request.QueryString), request.QueryString.Value);
+                }
+
+                if (_options.LoggingFields.HasFlag(HttpLoggingFields.RequestHeaders))
+                {
+                    FilterHeaders(list, request.Headers, _options.AllowedRequestHeaders);
+                }
+
+                if (_options.LoggingFields.HasFlag(HttpLoggingFields.RequestBody) && IsSupportedMediaType(request.ContentType))
+                {
+                    var body = await ReadRequestBody(request, context.RequestAborted);
+
+                    list.Add(new KeyValuePair<string, object?>(nameof(request.Body), body));
+                }    
+
+                // TODO add and remove things from log.
+                var httpRequestLog = new HttpRequestLog(list);
+
+                _logger.Log(LogLevel.Information,
+                     eventId: LoggerEventIds.RequestLog,
+                     state: httpRequestLog,
+                     exception: null,
+                     formatter: HttpRequestLog.Callback);
+            }
+
+            if ((HttpLoggingFields.Response & _options.LoggingFields) == HttpLoggingFields.None)
+            {
+                // Short circuit and don't replace response body.
+                await _next(context).ConfigureAwait(false);
+                return;
+            }
+
+            var response = context.Response;
+
+            ResponseBufferingStream? bufferingStream = null;
+            IHttpResponseBodyFeature? originalBodyFeature = null;
+            if (_options.LoggingFields.HasFlag(HttpLoggingFields.ResponseBody))
+            {
+                originalBodyFeature = context.Features.Get<IHttpResponseBodyFeature>()!;
+                // TODO pool these.
+                bufferingStream = new ResponseBufferingStream(originalBodyFeature, _options.ResponseBodyLogLimit);
+                response.Body = bufferingStream;
+            }
+
+            try
+            {
+                await _next(context).ConfigureAwait(false);
+                var list = new List<KeyValuePair<string, object?>>(response.Headers.Count + DefaultResponseFieldsMinusHeaders);
+
+                if (_options.LoggingFields.HasFlag(HttpLoggingFields.StatusCode))
+                {
+                    list.Add(new KeyValuePair<string, object?>(nameof(response.StatusCode), response.StatusCode));
+                }
+
+                if (_options.LoggingFields.HasFlag(HttpLoggingFields.ResponseHeaders))
+                {
+                    FilterHeaders(list, response.Headers, _options.AllowedResponseHeaders);
+                }
```

Comment:
The response log should be written on first write (or after next if there are no writes), and then the response body logged in chunks, not fully buffered.

This will have the added benefit of getting partial logs even if an exception is thrown part way through the response.